### ISSUE TITLE
pubsub: Test if a subscription exists in the Google Pub/Sub Go API using sub.Exists(ctx)

### ISSUE THREAD

**Issue by AfonsoVilalonga** [googleapis/google-cloud-go#9723]
Hello,

I was attempting to utilize the Google Pub/Sub Go API to create both a topic and a subscription to that topic. After creating both, I do a loop that loops until the sub.Exists(ctx) function returns true. As far as I understand, this function returns true if the subscription exists.

Once it returns true, I make a call to sub.Receive(ctx, func(ctx context.Context, msg *pubsub.Message)) to receive messages from the newly created topic using the newly created subscription.

However, sometimes (I cannot consistently replicate the error), I receive the following error from sub.Receive(ctx, func(ctx context.Context, msg *pubsub.Message)), even when sub.Exists(ctx) returns true:

rpc error: code = NotFound desc = Resource not found (resource=subscription_id).

Is this behavior normal?

**Comment by hongalex** [comment]
Are you checking if `CreateSubscription` returns an error in your code?

If so, this may be a bug with the Pub/Sub server, and not the Go client library. `sub.Exists` just makes a `GetSubscription` call to see if the subscription exists or not. If you are able to support ticket directly through Cloud Support, that would the best way to collect your project information to debug further.

**Comment by hongalex** [comment]
Could you share your `ReceiveSettings`? Theoretically, `StreamingPull` shouldn't ever return `NotFound` so I'm wondering if the error is coming from somewhere else in `Receive`.

**Comment by hongalex** [comment]
Is this still happening for you?

I just wanted to suggest that this pattern you are describing isn't really recommended for calling our API. Instead, we recommend optimistically assuming the subscription exists when calling `Receive`, and only creating a subscription if that returns false. [Please reference this sample](https://cloud.google.com/pubsub/docs/samples/pubsub-optimistic-subscribe#pubsub_optimistic_subscribe-go), with more info [here](https://cloud.google.com/pubsub/docs/troubleshooting#using_excessive_administrative_operations)

**Comment by AfonsoVilalonga** [comment]
Hi! Sorry for the lack of response on my end.

To give you some context, I was working on a system (for an academic paper) that manages a pool of users communicating with a server via the Google Pub/Sub service. The challenge was that I didn’t want users to see the content of messages the server was sending to other users. To address this, users could request the server to create a specific topic and its respective subscription on their behalf. This topic would then serve as a private communication channel between the user and the server because the names of the topic and the subscription would be known only by the server and the respective user. This is why the problem occurred—I needed the client to wait for the server to create the subscription and topic before sending messages.

I switched to an alternative approach that involves encrypting the content of the messages with symmetric keys established between each user and the server. This effectively solves the problem, so I stopped testing the other version of the system. I also assumed that this pattern was likely not recommended or intended for the API, but I found the behavior odd at the time, which is why I reported the issue.

I haven’t tested to see if the problem persists in a while, but now that I have some more time, I could test it again if you'd like.
